### PR TITLE
Move-away "uniqCombined" as a separate aggregated function

### PR DESCRIPTION
**Pull request by abyss7**
Accepts single param - the HLL precision.

### CODE REVIEW

File: dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    extern const int ARGUMENT_OUT_OF_BOUND;
+}
+
+namespace
+{
+
+AggregateFunctionPtr createAggregateFunctionUniqCombined(
+    const std::string & name, const DataTypes & argument_types, const Array & params)
+{
+    UInt8 precision = 17; /// default value - must correlate with default ctor of |AggregateFunctionUniqCombinedData|
```
Comment (4ertus2) [inline dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.cpp]: If it MUST be correlated it's better to set named constant and use it both here and in ctor.

File: dbms/tests/queries/0_stateless/00212_shard_aggregate_function_uniq.sql [hunk 2]
```diff
@@ -20,20 +20,20 @@ SELECT uniqHLL12(dummy) FROM remote('127.0.0.{2,3}', system.one);
 
 /* uniqCombined */
 
-SELECT Y, uniqCombined(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 15) GROUP BY Y;
-SELECT Y, uniqCombined(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 3000) GROUP BY Y;
-SELECT Y, uniqCombined(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 1000000) GROUP BY Y;
+SELECT Y, uniqCombined(15)(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 15) GROUP BY Y;
```
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00212_shard_aggregate_function_uniq.sql]: Would it change default behavior? What if someone already use uniqCombined without parameter. Would he (or she :) ) get the same values after update?

File: dbms/tests/queries/0_stateless/00212_shard_aggregate_function_uniq.sql [hunk 3]
```diff
@@ -20,20 +20,20 @@ SELECT uniqHLL12(dummy) FROM remote('127.0.0.{2,3}', system.one);
 
 /* uniqCombined */
 
-SELECT Y, uniqCombined(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 15) GROUP BY Y;
-SELECT Y, uniqCombined(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 3000) GROUP BY Y;
-SELECT Y, uniqCombined(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 1000000) GROUP BY Y;
+SELECT Y, uniqCombined(17)(X) FROM (SELECT number AS X, (3*X*X - 7*X + 11) % 37 AS Y FROM system.numbers LIMIT 15) GROUP BY Y;
```
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00212_shard_aggregate_function_uniq.sql]: What if someone already use uniqCombined without parameter? Would he or she get the same values after update?

File: dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.cpp (hunk 1 above)
Comment (abyss7) [reply dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.cpp]: It correlates with a hardcoded union member initialization, the constant won't help - so I left the comment. The code now looks like: `AggregateFunctionUniqCombinedDataWithKey() : set_17() {}` - Also I can use a macro-definition as constant, if it's better.

File: dbms/tests/queries/0_stateless/00212_shard_aggregate_function_uniq.sql (hunk 3 above)
Comment (abyss7) [reply dbms/tests/queries/0_stateless/00212_shard_aggregate_function_uniq.sql]: Yes, it had the value 17 - before my changes.

File: dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+} // namespace detail
+
+
+template <typename Key, UInt8 K>
+struct AggregateFunctionUniqCombinedDataWithKey
+{
+    using Set = CombinedCardinalityEstimator<Key,
+        HashSet<Key, TrivialHash, HashTableGrower<>>,
+        16,
+        K - 3,
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h]: Missing comment.

File: dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        K,
+        TrivialHash,
+        Key,
+        HyperLogLogBiasEstimator<UniqCombinedBiasData>,
```
Comment (alexey-milovidov) [inline dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h]: I'm not sure that it works correctly.

File: dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h (hunk 4 above)
Comment (alexey-milovidov) [reply dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h]: HyperLogLog uses 6 bits per cell.
HashSet uses `sizeof(Key) * 8 * 2` bits per element with maximum fill.

How `K - 3` is considered to be correct? Maybe it was already incorrect?

File: dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h (hunk 5 above)
Comment (abyss7) [reply dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h]: Will generate new data for all values of K in range [12,20].

Comment (abyss7) [reply dbms/src/AggregateFunctions/AggregateFunctionUniqCombined.h]: Actually decided not to do this right now.

**Comment by alexey-milovidov** [comment]
https://gist.github.com/alexey-milovidov/9cbbc2916f8d946594049d424487e458

Now the `uniqHLL12` function can be considered obsolete.

**Pull request merged**